# The Battle of Neighborhoods

### Instalando dependencias

In [5]:
!conda install -c conda-forge geopy --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [3]:
!conda install -c conda-forge folium --yes

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1f             |       h516909a_0         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    certifi-2020.4.5.1         |   py36h9f0ad1d_0         151 KB  conda-forge
    ca-certificates-2020.4.5.1 |       hecc5488_0         146 KB  conda-forge
    branca-0.4.0               |             py_0          26 KB  conda-forge
    folium-0.10.1              |             py_0          59 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTAL

### Import packages

In [4]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm # Matplotlib and associated plotting modules
import matplotlib.colors as colors
from sklearn.cluster import KMeans # import k-means from clustering stage
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Download and Explore Dataset

#### Dataset 1. Coordinates of the neighborhoods
Obtaining coordinate data for New York neighborhoods

In [5]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [6]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Exploring the data structure and obtaining the necessary data

In [8]:
newyork_data['features'][0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [9]:
neighborhoods_data = newyork_data['features']

In [10]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

"neighborhoods" will contain the coordinate data

In [11]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Loading the source data to "neighborhoods" target dataframe

In [12]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [13]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [14]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


The dataframe "neighborhoods" has 5 boroughs and 306 neighborhoods. According to the data requirements, the neighborhoods of Manhattan will be considered, in this way we obtain the data set of Manhattan

In [15]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

#### Dataset 2. Foursquare API to get and explore Manhattan venues

In [16]:
CLIENT_ID = 'N1QOETOCEDFTSUI2GA0OFHVDAWS5XXHDJ4JWJENBYXVDY2ZI' # your Foursquare ID
CLIENT_SECRET = 'ZAJVVTHMNHXTDHESFDHMXPM3UW5UELOY4IFPLVNQ2YH0RMXV' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: N1QOETOCEDFTSUI2GA0OFHVDAWS5XXHDJ4JWJENBYXVDY2ZI
CLIENT_SECRET:ZAJVVTHMNHXTDHESFDHMXPM3UW5UELOY4IFPLVNQ2YH0RMXV


Let's getting 100 places 500 meters around for each neighborhood

In [19]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT = 100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


"manhattan_venues" dataset contains the data of the neighborhoods and their places located 500 meters around with their respective coordinates. Exploring the resulting dataset "manhattan_venues"

In [22]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3158, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066        Dunkin'   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877136       -73.906666     Donut Shop  
4       40.877531       -73.905582    Coffee Shop

### Analysis and Preparing Dataset

For our analysis we must prepare the dataset for our model, therefore the categorical variable "venues" we will apply the one hot encoding process

In [23]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  ...  Video Store  \
0                 0                       0  ...            0   
1                 0                       0  ...            0   
2                 0                       0  ...            0   
3                 0                       0  ...            0   
4                 0                       0  ...            0   

   Vietnamese Restaurant  Volleyball Court  Waterfront  Whisky Bar  Wine Bar  \
0                      0                 0           0           0         0   
1                      0                 0           0           0         0   
2                      0                 0           0           0         0   
3                      0                 0           0           0         0   
4                      0                 0           0           0         0   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0          0            0              0            0  
1          0            0              0            1  
2          0            0              0            0  
3          0            0              0            0  
4          0            0              0            0  

[5 rows x 324 columns]

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [24]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.000000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.010000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.00               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.000000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.000000            0.00               0.00   
37   Washington Heights           0.011364            0.00               0.00   
38         West Village           0.000000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Antique Shop    Arcade  \
0             0.000000             0.014493          0.00  0.000000   
1             0.000000             0.011364          0.00  0.000000   
2             0.044444             0.044444          0.00  0.000000   
3             0.000000             0.040000          0.00  0.000000   
4             0.000000             0.040000          0.00  0.000000   
5             0.000000             0.030000          0.01  0.000000   
6             0.000000             0.030000          0.00  0.000000   
7             0.000000             0.000000          0.00  0.000000   
8             0.000000             0.020000       

Look at each neighborhood along with the top 5 most common venues

In [25]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0           Park  0.09
1          Hotel  0.07
2            Gym  0.06
3    Coffee Shop  0.04
4  Memorial Site  0.04


----Carnegie Hill----
                 venue  freq
0          Coffee Shop  0.08
1                 Café  0.06
2  Japanese Restaurant  0.03
3               Bakery  0.03
4   Italian Restaurant  0.03


----Central Harlem----
                  venue  freq
0    Chinese Restaurant  0.04
1  Gym / Fitness Center  0.04
2    Seafood Restaurant  0.04
3     French Restaurant  0.04
4   Fried Chicken Joint  0.04


----Chelsea----
                 venue  freq
0          Coffee Shop  0.09
1          Art Gallery  0.05
2  American Restaurant  0.04
3   Italian Restaurant  0.03
4               Bakery  0.03


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.09
1         Cocktail Bar  0.05
2  American Restaurant  0.04
3               Bakery  0.04
4                  Spa  0.03


----Civic Center----
               venue  fr

Finally we remove the "Neighborhood" column to get our input dataset to the model.

In [26]:
manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)
manhattan_grouped_clustering.head()

Accessories Store  Adult Boutique  Afghan Restaurant  African Restaurant  \
0                0.0             0.0                0.0            0.000000   
1                0.0             0.0                0.0            0.000000   
2                0.0             0.0                0.0            0.044444   
3                0.0             0.0                0.0            0.000000   
4                0.0             0.0                0.0            0.000000   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0             0.014493           0.0     0.0               0.0   
1             0.011364           0.0     0.0               0.0   
2             0.044444           0.0     0.0               0.0   
3             0.040000           0.0     0.0               0.0   
4             0.040000           0.0     0.0               0.0   

   Argentinian Restaurant  Art Gallery  ...  Video Store  \
0                     0.0     0.000000  ...          0.0   
1                     0.0     0.000000  ...          0.0   
2                     0.0     0.044444  ...          0.0   
3                     0.0     0.050000  ...          0.0   
4                     0.0     0.000000  ...          0.0   

   Vietnamese Restaurant  Volleyball Court  Waterfront  Whisky Bar  Wine Bar  \
0               0.000000               0.0         0.0         0.0  0.000000   
1               0.011364               0.0         0.0         0.0  0.011364   
2               0.000000               0.0         0.0         0.0  0.000000   
3               0.000000               0.0         0.0         0.0  0.010000   
4               0.030000               0.0         0.0         0.0  0.000000   

   Wine Shop  Wings Joint  Women's Store  Yoga Studio  
0   0.028986          0.0       0.014493     0.000000  
1   0.034091          0.0       0.011364     0.034091  
2   0.000000          0.0       0.000000     0.000000  
3   0.010000          0.0       0.000000     0.010000  
4   0.000000          0.0       0.000000     0.010000  

[5 rows x 323 columns]

### Modeling

In the business understanding the analytical approach of the data was determined, this is the K means clustering algorithm

In [27]:
# set number of clusters
kclusters = 5

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 2, 3, 2, 3, 2, 2, 3, 2, 1], dtype=int32)

### Results

The generated labels will be associated to the corresponding neighborhood to observe the clusters. For this we will create a DataFrame with the neighborhoods and the 10 most frequent places that we can find around

In [28]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [29]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park                 Hotel   
1      Carnegie Hill           Coffee Shop                  Café   
2     Central Harlem  Gym / Fitness Center     French Restaurant   
3            Chelsea           Coffee Shop           Art Gallery   
4          Chinatown    Chinese Restaurant          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Gym           Coffee Shop         Memorial Site   
1           Yoga Studio    Italian Restaurant   Japanese Restaurant   
2   Fried Chicken Joint    African Restaurant   American Restaurant   
3   American Restaurant                Bakery    Italian Restaurant   
4   American Restaurant                Bakery                   Spa   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Shopping Mall          Gourmet Shop             Wine Shop   
1                Bakery                   Gym  Gym / Fitness Center   
2                   Bar    Seafood Restaurant    Chinese Restaurant   
3  Gym / Fitness Center          Cocktail Bar           Pizza Place   
4          Dessert Shop          Optical Shop     Hotpot Restaurant   

   9th Most Common Venue 10th Most Common Venue  
0     Mexican Restaurant             Food Court  
1            Pizza Place         Cosmetics Shop  
2            Art Gallery              Bookstore  
3              Pet Store              Nightclub  
4  Vietnamese Restaurant     Salon / Barbershop

Labeling neighborhoods (Cluster Labels)

In [30]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               1   
1  Manhattan           Chinatown  40.715618 -73.994279               3   
2  Manhattan  Washington Heights  40.851903 -73.936900               3   
3  Manhattan              Inwood  40.867684 -73.921210               3   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0        Sandwich Place           Coffee Shop                   Gym   
1    Chinese Restaurant          Cocktail Bar   American Restaurant   
2                  Café                Bakery         Grocery Store   
3    Mexican Restaurant           Pizza Place            Restaurant   
4           Pizza Place                  Café           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0        Ice Cream Shop      Department Store              Pharmacy   
1                Bakery                   Spa          Dessert Shop   
2    Chinese Restaurant     Mobile Phone Shop         Deli / Bodega   
3                  Café                Lounge    Frozen Yogurt Shop   
4         Deli / Bodega    Mexican Restaurant                Bakery   

  7th Most Common Venue 8th Most Common Venue  9th Most Common Venue  \
0                 Diner        Discount Store             Donut Shop   
1          Optical Shop     Hotpot Restaurant  Vietnamese Restaurant   
2           Coffee Shop           Supermarket        Supplement Shop   
3                Bakery                  Park     Chinese Restaurant   
4      Sushi Restaurant          Cocktail Bar         Sandwich Place   

  10th Most Common Venue  
0                   Bank  
1     Salon / Barbershop  
2       Tapas Restaurant  
3          Deli / Bodega  
4            Yoga Studio

Let's visualize the resulting clusters in the map, therefore we get the geographical coordinates of Manhattan.

In [32]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


In [33]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Finally, let's visualize the resulting clusters

### Cluster 1

In [34]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
33  Midtown South     Korean Restaurant                 Hotel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
33   Japanese Restaurant          Dessert Shop          Burger Joint   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
33  Gym / Fitness Center        Clothing Store                   Bar   

   9th Most Common Venue 10th Most Common Venue  
33   American Restaurant            Coffee Shop

### Cluster 2

In [36]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0           Marble Hill        Sandwich Place           Coffee Shop   
11     Roosevelt Island                  Park           Pizza Place   
15              Midtown           Coffee Shop                 Hotel   
16          Murray Hill        Sandwich Place                 Hotel   
26  Morningside Heights                  Park             Bookstore   
28    Battery Park City                  Park                 Hotel   
29   Financial District           Coffee Shop           Pizza Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                    Gym        Ice Cream Shop      Department Store   
11        Sandwich Place  Gym / Fitness Center                   Gym   
15   Sporting Goods Shop        Clothing Store            Steakhouse   
16           Coffee Shop      Sushi Restaurant                   Bar   
26   American Restaurant           Coffee Shop         Deli / Bodega   
28                   Gym           Coffee Shop         Memorial Site   
29                 Hotel    Mexican Restaurant          Cocktail Bar   

   6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0               Pharmacy                 Diner         Discount Store   
11           Coffee Shop          Liquor Store  Outdoors & Recreation   
15                   Spa                Bakery                   Café   
16   Japanese Restaurant          Burger Joint   Gym / Fitness Center   
26          Burger Joint                   Pub         Farmers Market   
28         Shopping Mall          Gourmet Shop              Wine Shop   
29                   Bar                  Park                    Gym   

       9th Most Common Venue 10th Most Common Venue  
0                 Donut Shop                   Bank  
11               Dry Cleaner                 School  
15            Sandwich Place                Theater  
16       American Restaurant       Cuban Restaurant  
26  Mediterranean Restaurant            Supermarket  
28        Mexican Restaurant             Food Court  
29      Gym / Fitness Center              Juice Bar

### Cluster 3

In [37]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
5      Manhattanville           Coffee Shop       Seafood Restaurant   
8     Upper East Side    Italian Restaurant              Coffee Shop   
9           Yorkville    Italian Restaurant              Coffee Shop   
10         Lenox Hill    Italian Restaurant              Coffee Shop   
12    Upper West Side    Italian Restaurant              Coffee Shop   
13     Lincoln Square                 Plaza                     Café   
14            Clinton               Theater     Gym / Fitness Center   
17            Chelsea           Coffee Shop              Art Gallery   
18  Greenwich Village    Italian Restaurant                     Café   
19       East Village                   Bar              Pizza Place   
21            Tribeca                  Park       Italian Restaurant   
23               Soho           Coffee Shop       Italian Restaurant   
24       West Village    Italian Restaurant  New American Restaurant   
25   Manhattan Valley           Coffee Shop                      Bar   
27           Gramercy                   Bar       Italian Restaurant   
30      Carnegie Hill           Coffee Shop                     Café   
31               Noho    Italian Restaurant              Coffee Shop   
32       Civic Center           Coffee Shop        French Restaurant   
34       Sutton Place           Coffee Shop       Italian Restaurant   
35         Turtle Bay    Italian Restaurant              Coffee Shop   
38           Flatiron  Gym / Fitness Center  New American Restaurant   
39       Hudson Yards                 Hotel     Gym / Fitness Center   

   3rd Most Common Venue  4th Most Common Venue  \
5           Liquor Store     Italian Restaurant   
8                 Bakery                Exhibit   
9                    Gym                    Bar   
10      Sushi Restaurant            Pizza Place   
12                   Bar                 Bakery   
13  Gym / Fitness Center  Performing Arts Venue   
14           Coffee Shop     Italian Restaurant   
17   American Restaurant                 Bakery   
18           Coffee Shop                    Gym   
19              Wine Bar            Coffee Shop   
21                   Spa                   Café   
23                  Café                 Bakery   
24   American Restaurant               Wine Bar   
25     Indian Restaurant     Mexican Restaurant   
27   American Restaurant            Coffee Shop   
30           Yoga Studio     Italian Restaurant   
31                 Hotel            Pizza Place   
32          Cocktail Bar                  Hotel   
34  Gym / Fitness Center                    Gym   
35                 Hotel          Deli / Bodega   
38                  Café            Yoga Studio   
39   American Restaurant     Italian Restaurant   

            5th Most Common Venue   6th Most Common Venue  \
5                   Deli / Bodega                    Park   
8                             Spa       French Restaurant   
9                   Deli / Bodega             Pizza Place   
10                   Cocktail Bar                    Café   
12                    Pizza Place          Ice Cream Shop   
13                        Theater            Concert Hall   
14                            Spa         Thai Restaurant   
17             Italian Restaurant    Gym / Fitness Center   
18              French Restaurant            Dessert Shop   
19             Mexican Restaurant               Speakeasy   
21                       Wine Bar             Coffee Shop   
23                 Sandwich Place          Clothing Store   
24                         Bakery            Cocktail Bar   
25                Thai Restaurant             Yoga Studio   
27                    Pizza Place           Grocery Store   
30            Japanese Restaurant                  Bakery   
31            Japanese Restaurant       French Restaurant   
32                    Yoga Studio                    Café   
34                 

### Cluster 4

In [38]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
1            Chinatown    Chinese Restaurant          Cocktail Bar   
2   Washington Heights                  Café                Bakery   
3               Inwood    Mexican Restaurant           Pizza Place   
4     Hamilton Heights           Pizza Place                  Café   
6       Central Harlem  Gym / Fitness Center     French Restaurant   
7          East Harlem    Mexican Restaurant                Bakery   
20     Lower East Side    Chinese Restaurant           Art Gallery   
22        Little Italy                Bakery                  Café   
36          Tudor City                  Park                  Café   

   3rd Most Common Venue      4th Most Common Venue     5th Most Common Venue  \
1    American Restaurant                     Bakery                       Spa   
2          Grocery Store         Chinese Restaurant         Mobile Phone Shop   
3             Restaurant                       Café                    Lounge   
4            Coffee Shop              Deli / Bodega        Mexican Restaurant   
6    Fried Chicken Joint         African Restaurant       American Restaurant   
7        Thai Restaurant  Latin American Restaurant             Deli / Bodega   
20   Japanese Restaurant                     Bakery               Pizza Place   
22    Chinese Restaurant            Bubble Tea Shop  Mediterranean Restaurant   
36         Deli / Bodega         Mexican Restaurant                     Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1           Dessert Shop          Optical Shop     Hotpot Restaurant   
2          Deli / Bodega           Coffee Shop           Supermarket   
3     Frozen Yogurt Shop                Bakery                  Park   
4                 Bakery      Sushi Restaurant          Cocktail Bar   
6                    Bar    Seafood Restaurant    Chinese Restaurant   
7            Gas Station          Cocktail Bar          Liquor Store   
20                  Café      Ramen Restaurant           Coffee Shop   
22                 Hotel     Hotpot Restaurant    Italian Restaurant   
36      Greek Restaurant                Garden       Thai Restaurant   

    9th Most Common Venue 10th Most Common Venue  
1   Vietnamese Restaurant     Salon / Barbershop  
2         Supplement Shop       Tapas Restaurant  
3      Chinese Restaurant          Deli / Bodega  
4          Sandwich Place            Yoga Studio  
6             Art Gallery              Bookstore  
7              Taco Place         Sandwich Place  
20            Yoga Studio        Bubble Tea Shop  
22        Thai Restaurant            Coffee Shop  
36   Gym / Fitness Center       Sushi Restaurant

### Cluster 5

In [39]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town         Boat or Ferry                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37                   Bar           Pet Service     German Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37        Baseball Field              Fountain       Harbor / Marina   

   9th Most Common Venue 10th Most Common Venue  
37          Cocktail Bar            Coffee Shop